File: src/main/java/fi/metatavu/metaform/server/persistence/model/Metaform.java
package fi.metatavu.metaform.server.persistence.model;

import java.util.UUID;

import javax.persistence.Cacheable;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Lob;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

import org.hibernate.annotations.Cache;
import org.hibernate.annotations.CacheConcurrencyStrategy;

/**
 * JPA entity representing single Metaform
 * 
 * @author Antti Leppä
 */
@Entity
@Cacheable(true)
@Cache(usage = CacheConcurrencyStrategy.TRANSACTIONAL)
@Table (
  uniqueConstraints = {
    @UniqueConstraint (columnNames = { "slug" })  
  }    
)
public class Metaform {

  @Id
  private UUID id;

  @NotNull
  @NotEmpty
  private String slug;

  @Lob
  private String data;

  @NotNull
  @Column (nullable = false)
  private Boolean allowAnonymous;
  
  @ManyToOne
  private ExportTheme exportTheme;
  
  public UUID getId() {
    return id;
  }
  
  public void setId(UUID id) {
    this.id = id;
  }
  
  public String getSlug() {
    return slug;
  }
  
  public void setSlug(String slug) {
    this.slug = slug;
  }
  
  public String getData() {
    return data;
  }
  
  public void setData(String data) {
    this.data = data;
  }
  
  public Boolean getAllowAnonymous() {
    return allowAnonymous;
  }
  
  public void setAllowAnonymous(Boolean allowAnonymous) {
    this.allowAnonymous = allowAnonymous;
  }
  
  public ExportTheme getExportTheme() {
    return exportTheme;
  }
  
  public void setExportTheme(ExportTheme exportTheme) {
    this.exportTheme = exportTheme;
  }
  
}

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/Metaform.kt
package fi.metatavu.metaform.server.persistence.model

import fi.metatavu.metaform.api.spec.model.MetaformVisibility
import org.hibernate.annotations.Cache
import org.hibernate.annotations.CacheConcurrencyStrategy
import java.time.OffsetDateTime
import java.util.*
import jakarta.persistence.*
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.NotNull

/**
 * JPA entity representing single Metaform
 *
 * @author Antti Leppä
 */
@Entity
@Cacheable(true)
@Cache(usage = CacheConcurrencyStrategy.TRANSACTIONAL)
@Table(uniqueConstraints = [UniqueConstraint(columnNames = ["slug"])])
class Metaform: Metadata() {
  @Id
  var id: UUID? = null

  @Column(nullable = false)
  @NotNull
  @NotEmpty
  lateinit var slug: String

  @Lob
  @Column(nullable = false)
  @NotNull
  @NotEmpty
  lateinit var data: String

  @Column(nullable = false)
  @NotNull
  var allowAnonymous: Boolean? = null

  @NotNull
  @Enumerated(EnumType.STRING)
  var visibility: MetaformVisibility? = null

  @ManyToOne
  var exportTheme: ExportTheme? = null

  @Column(nullable = false)
  lateinit var creatorId: UUID

  @Column(nullable = false)
  lateinit var lastModifierId: UUID

  @Column(nullable = false)
  @NotNull
  var active: Boolean = true

  @OneToMany(mappedBy = "metaform", targetEntity = MetaformReplyViewed::class)
  lateinit var metaformReplyViewed: List<MetaformReplyViewed>

  @OneToMany(mappedBy = "metaform", targetEntity = MetaformReplyCreated::class)
  lateinit var metaformReplyCreated: List<MetaformReplyCreated>
}